storage/cloudspanner/tree_storage.go

```diff
@@ -0,0 +1,523 @@
+	// with a read-only transaction.  This should not even be possible.
+	ErrWrongTXType = errors.New("mutating method called on read-only transaction")
+
+	// errFinished is only used to terminate spanner reads early, once all
+	// required data has been read. It should never be returned to a caller.
+	errFinished = errors.New("read complete")
+)
+
+const (
+	subtreeTbl   = "SubtreeData"
+	colBucket    = "Bucket"
+	colSubtree   = "Subtree"
+	colSubtreeID = "SubtreeID"
+	colTreeID    = "TreeID"
+	colRevision  = "Revision"
+)
+
+// treeStorage provides a shared base for the concrete Spanner-backed
+// implementation of the Trillian storage.LogStorage and storage.MapStorage
+// interfaces.
+type treeStorage struct {
+	admin  storage.AdminStorage
+	opts   TreeStorageOptions
+	client *spanner.Client
+}
+
+// TreeStorageOptions holds various levers for configuring the tree storage instance.
+type TreeStorageOptions struct {
+	// ReadOnlyStaleness controls how far in the past a read-only snapshot
+	// transaction will read.
+	// This is intended to allow Spanner to use local replicas for read requests
+	// to help with performance.
+	ReadOnlyStaleness time.Duration
+}
+
+func newTreeStorageWithOpts(client *spanner.Client, opts TreeStorageOptions) *treeStorage {
+	return &treeStorage{client: client, admin: nil, opts: opts}
+}
+
+type spanRead interface {
+	Query(context.Context, spanner.Statement) *spanner.RowIterator
+	Read(ctx context.Context, table string, keys spanner.KeySet, columns []string) *spanner.RowIterator
+	ReadUsingIndex(ctx context.Context, table, index string, keys spanner.KeySet, columns []string) *spanner.RowIterator
+	ReadRow(ctx context.Context, table string, key spanner.Key, columns []string) (*spanner.Row, error)
+}
+
+// latestSTH reads and returns the newest STH.
+func (t *treeStorage) latestSTH(ctx context.Context, stx spanRead, treeID int64) (*spannerpb.TreeHead, error) {
+	query := spanner.NewStatement(
+		"SELECT t.TreeID, t.TimestampNanos, t.TreeSize, t.RootHash, t.RootSignature, t.TreeRevision, t.TreeMetadata FROM TreeHeads t" +
+			"   WHERE t.TreeID = @tree_id" +
+			"   ORDER BY t.TimestampNanos DESC " +
+			"   LIMIT 1")
+	query.Params["tree_id"] = treeID
+
+	var th *spannerpb.TreeHead
+	rows := stx.Query(ctx, query)
+	defer rows.Stop()
+	err := rows.Do(func(r *spanner.Row) error {
+		tth := &spannerpb.TreeHead{}
+		var sig, meta []byte
+		if err := r.Columns(&tth.TreeId, &tth.TsNanos, &tth.TreeSize, &tth.RootHash, &sig, &tth.TreeRevision, &meta); err != nil {
+			return err
+		}
+		sigPB := &spannerpb.DigitallySigned{}
+		if err := proto.Unmarshal(sig, sigPB); err != nil {
+			return err
+		}
+		tth.Signature = sigPB
+
+		metaPB := &any.Any{}
+		if err := proto.Unmarshal(meta, metaPB); err != nil {
+			return err
+		}
+		tth.Metadata = metaPB
+
+		th = tth
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	if th == nil {
+		glog.Warningf("no head found for treeID %v", treeID)
```

Comment:
Maybe info rather than warning? It's not unexpected.

Reply:
But if there are many of these turning up then it's a problem somewhere...

Reply:
OK it looks like glog doesn't have the LOG_EVERY_.... stuff from C++.